### PR TITLE
update user settings - issue 20

### PR DESCRIPTION
**Pull request by JoeArmani**
Link to issue: [https://github.com/episphere/connectApp/issues/20](url)

This PR addresses connectApp issue #20 - allow members to update their profile information once they’re verified.

**Functionality:**

Profile must be verified to be editable.

- If submitted, user can see their data. If verified, user can edit their data.

Refactored HTML for readability, separation of concerns, and faster dev access

Integrated conceptId mapping for readability.

Created toggling edit & submit buttons for name, contact info, and mailing address.

Updated Name display to include additional fields based on user’s data availability.

Created Name Form.

- 	dynamically load fields based on user’s information.
- 	validate client-side.
- 	integrated data submit.

Updated Contact Information display to include additional fields based on user’s data.

Created Contact information form.

- 	dynamically load fields based on user’s information.
- 	validate client-side.
- 	integrated data submit.

Created Mailing address update form.

- 	uses Google Maps API for easy and correct address data.
- 	validates client-side.
- 	integrates data submit.

Updated Sign-In Information to match code stylistically, so settings.js and settingsHelper.js are cohesive moving forward.

Notes based on original issue:

- Salutation conceptId not found
- WhatsApp conceptId not found

### CODE REVIEW

File: js/pages/settings.js [hunk 1]
```diff
@@ -1,327 +1,996 @@
...
+  let canWeText = userData[cId.canWeText] === cId.yes;
+  let canWeVoicemailHome = userData[cId.canWeVoicemailHome] === cId.yes;
+  middleName = userData[cId.mName];
+  suffix = userData[cId.suffix];
+  preferredFirstName = userData[cId.prefName];
+  otherPhoneNumberComplete = userData[cId.otherPhone];
+  additionalEmail = userData[cId.additionalEmail];
+  let template = '';
+
+  if (myData.code === 200 && userData[cId.userProfileSubmittedAutogen] !== undefined && userData[cId.userProfileSubmittedAutogen] === cId.yes) {
```
Comment (we-ai) [inline js/pages/settings.js:27]:
```suggestion
  if (myData.code === 200 && userData[cId.userProfileSubmittedAutogen] === cId.yes) {
```

File: js/pages/settings.js [hunk 2]
```diff
@@ -1,327 +1,996 @@
...
+   * If the user profile has been verified, then show the profile and edit functionality:
+   * Create the buttons, add the event listeners, push the elements to the arrays for visibility toggling
+   * If the user profile has not been verified, then hide the profile and edit functionality, show the pending verification message
+   */
+  if (
+    myData.code === 200 &&
+    userData[cId.userProfileSubmittedAutogen] !== undefined &&
+    userData[cId.userProfileSubmittedAutogen] === cId.yes &&
+    userData[cId.verified] !== undefined &&
+    userData[cId.verified] === cId.yes
```
Comment (we-ai) [inline js/pages/settings.js:93]:
```suggestion
    userData[cId.userProfileSubmittedAutogen] === cId.yes &&
    userData[cId.verified] === cId.yes
```

File: js/settingsHelpers.js [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+};
+
+export const FormTypes = {
+  NAME: 'nameForm',
+  CONTACT: 'contactForm',
+  MAILING: 'mailingForm',
+  EMAIL: 'emailForm',
+};
+
+export const validEmailFormat = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
```
Comment (we-ai) [inline js/settingsHelpers.js:89]: We have a `validEmailFormat` in `shared.js` file.

File: js/pages/settings.js [hunk 4]
```diff
@@ -1,327 +1,996 @@
...
-    let template = ''
-    
-    if(myData.code === 200 && myData.data['699625233'] !== undefined && myData.data['699625233'] === 353358909){
-        const userData = myData.data;
-
-        template += `
+  document.title = 'My Connect - My Profile';
+  showAnimation();
+  const myData = await getMyData();
+  const userData = myData.data;
```
Comment (jhflorey) [inline js/pages/settings.js:16]: Should we check the data exists

File: js/settingsHelpers.js [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+
+export const updatePhoneNumberInputFocus = () => {
```
Comment (jhflorey) [inline js/settingsHelpers.js:91]: Should we create common function like

const focusEvent = (id) => {
const element = document.getElementById(id);
  element.addEventListener('keyup', () => {
    if (element.value.trim().length === 3) {
      element.nextElementSibling.nextElementSibling.focus();
    }
  });
}

File: js/settingsHelpers.js (hunk 3 above)
Comment (we-ai) [reply js/settingsHelpers.js:89]: `validEmailFormat` in `shared.js` is not enforcing top level domains. I can update it. There's a related spec about email format [here](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address)

File: js/settingsHelpers.js (hunk 5 above)
Comment (JoeArmani) [reply js/settingsHelpers.js:91]: Great suggestion. It's only called on update contact information, so I'll just iterate the array.

`
export const updatePhoneNumberInputFocus = () => {
  const phoneElementIds = ['mobilePhoneNumber1', 'mobilePhoneNumber2', 'homePhoneNumber1', 'homePhoneNumber2', 'otherPhoneNumber1', 'otherPhoneNumber2'];

  const initFocusHandler = elementId => {
    const element = document.getElementById(elementId);
    element.addEventListener('keyup', () => {
      if (element.value.trim().length === 3) {
        element.nextElementSibling.nextElementSibling.focus();
      }
    });
  };

  phoneElementIds.forEach(initFocusHandler);
};
`

File: js/pages/settings.js (hunk 4 above)
Comment (JoeArmani) [reply js/pages/settings.js:16]: On one hand the user doc is written by cloud functions and the user is authenticated at this point in the app flow. Chances are almost nonexistent that there won't be a user doc. On the other hand chances are probably not zero - it seems like a worthwhile check - I agree with you and I'll add this. Thanks 👍

File: js/settingsHelpers.js (hunk 3 above)
Comment (JoeArmani) [reply js/settingsHelpers.js]: resolved with regexp in `shared.js` that enforces TLD (min 2-char). It validates Unicode and Punycode-encoded Unicode.

**Comment by JoeArmani** [comment]
Updates in 2nd commit:
-resolved all comments/suggestions to date
-updated `getMyData()` in `shared.js` to include error handling.
-move `handler functions` outside of `renderSettingsPage()` in `settings.js`
-added check if data exists after `getMyData()` call. 
-updated variable scoping and page loading logic to accommodate above changes.

**Comment by JoeArmani** [comment]
Updates in 3rd commit:
-resolved phone number flexibility to mimic sign up flow: user can enter mobile, home, or other phone number. Only one is required. User has the option to enter two or all three, now the UX plays nicely with those options.
-expanded upon name validation to include some Latin characters and special characters.
-updated edit form styling for readability
-added `canWeVoicemail` field for `otherPhoneNumber`

**Comment by JoeArmani** [comment]
Final commit:
-enforce whatwg.org email validation standard + TLD requirement.
-clean up style.css

**Pull request merged**